sdk/cosmosdb/cosmos/src/retry/timeoutFailoverRetryPolicy.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { RetryPolicy } from "./RetryPolicy";
+import { StatusCodes } from "../common/statusCodes";
+import { GlobalEndpointManager } from "../globalEndpointManager";
+import { HTTPMethod, isReadRequest } from "../common";
+import { ErrorResponse } from "../request";
+import { Constants, OperationType, ResourceType } from "../common/constants";
+import { RetryContext } from "./RetryContext";
+import { CosmosHeaders } from "../queryExecutionContext/CosmosHeaders";
+import { TimeoutErrorCode } from "../request/TimeoutError";
+
+/**
+ * This class TimeoutFailoverRetryPolicy handles retries for read operations
+ * (including data plane,metadata, and query plan) in case of request timeouts
+ * (TimeoutError) or service unavailability (503 status code) by performing failover
+ * and retrying on other regions.
+ * @hidden
+ */
+export class TimeoutFailoverRetryPolicy implements RetryPolicy {
+  private maxRetryAttemptCount = 120;
+  private maxServiceUnavailableRetryCount = 1;
+  public retryAfterInMs = 0;
+  public failoverRetryCount = 0;
+  public request: any;
+  public locationEndpoint: any;
+
+  constructor(
+    private globalEndpointManager: GlobalEndpointManager,
+    private headers: CosmosHeaders,
+    private methodType: HTTPMethod,
+    private resourceType: ResourceType,
+    private operationType: OperationType,
+    private enableEndPointDiscovery: boolean
+  ) {}
+
+  /**
+   * Checks if a timeout request is valid for the timeout failover retry policy.
+   * A valid request should be a data plane, metadata, or query plan request.
+   * @returns
+   */
+  private isValidRequestForTimeoutError(): boolean {
+    const isQuery = Constants.HttpHeaders.IsQuery in this.headers;
```

Comment:
Just want to explicitly make sure this - These headers are client-side headers, not returned from server. Because this policy is initialized before the request is actually made to server. This will work if client always sets this header, Are we sure the client always sets these headers. I am sure you have made sure this, Just asking.

Reply:
Yup, We do set these fields within client.